rpki-validator/src/main/java/net/ripe/rpki/validator3/background/Throttled.java

```diff
@@ -44,85 +43,104 @@
  * The purpose of this class is to be able to execute an arbitrary action
  * not more often than a certain interval. I.e.
  * <p>
- * val throttled = new Throttled(10)
+ * val throttled = new Throttled(10_000)
  * <p>
- * throttled.trigger("action1", r) - will execute immediately.
+ * throttled.trigger("action1", r1) - will execute r1 immediately.
  * <p>
- * Thread.sleep(3)
+ * Thread.sleep(3000)
  * <p>
- * throttled.trigger("action1", r) - will schedule execution in 7 (i.e. 10 - 3) seconds from now.
- * throttled.trigger("action1", r) - will be ignored as there's already execution scheduled.
+ * throttled.trigger("action1", r2) - will schedule execution of r2 in 7 (i.e. 10 - 3) seconds from now.
+ * throttled.trigger("action1", r3) - will replace scheduled execution with r3.
  * <p>
- * Thread.sleep(10)
+ * Thread.sleep(7010);
+ * throttled.trigger("action1", r4) - executes r4 after the currently running action completes ensuring the minimum delay.
  * <p>
- * throttled.trigger("action1", r) - will execute immediately again.
+ * Thread.sleep(11000)
+ * <p>
+ * throttled.trigger("action1", r5) - will execute r5 immediately again.
  */
 public class Throttled<Key> {
 
-    private final long minIntervalInSeconds;
+    private final long minIntervalMs;
 
     private final Map<Key, Action> actionMap = new HashMap<>();
 
     private final ScheduledExecutorService scheduledExecutor =
         Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
 
-    public Throttled(long minIntervalInSeconds) {
-        this.minIntervalInSeconds = minIntervalInSeconds;
+    public Throttled(long minIntervalMs) {
+        this.minIntervalMs = minIntervalMs;
     }
 
     public void trigger(Key key, Runnable r) {
-        final Runnable wrappedRunnable = () -> {
+        synchronized (actionMap) {
+            final Instant now = Instant.now();
+            final Action action = actionMap.get(key);
+
+            if (action == null) {
+                actionMap.put(key, Action.toBeExecutedASAP(now, r));
+                scheduledExecutor.execute(actionRunner(key));
+            } else if (action.running) {
+                actionMap.put(key, action.scheduled(r));
+            } else if (action.alreadyScheduled) {
+                actionMap.put(key, action.replaced(r));
+            } else {
+                final Instant lastTime = action.getExecutionTime();
+                final long delay = minIntervalMs - (now.toEpochMilli() - lastTime.toEpochMilli());
```

Comment:
Should we handle negative delay case? Can it ever happen?

Reply:
Negative delay results in an immediate execution.